### ISSUE TITLE
Can Flat decorate a structure object?

### ISSUE THREAD

**Issue by johnkslang** [KhronosGroup/SPIRV-Headers#17]
This comes from https://github.com/KhronosGroup/glslang/issues/418.  I'm copying relevant spec. text here:

From the SPIR-V specification:

> When applied to structure-type members, the Decorations Noperspective, **Flat**, Patch, Centroid, and Sample can only be applied to the top-level members of the structure type. (Nested objects' types cannot be structures whose members are decorated with these decorations.)

Also:

> **Flat**
>
> Apply to an object or a member of a structure type. Indicates no interpolation will be done. The non-interpolated value will come from a vertex, as described in the API specification. The object or member must be a scalar or vector of floating-point type or integer type. Arrays of these types are also allowed. Only valid for the Input and Output Storage Classes.

(This probably came from before deep inheritance propagation was disallowed in SPIR-V.  When we decided to have it apply to top-level members only, this rule should have allowed structures.)

So, we have one rule saying it can only apply to top-level members, and another rule saying it can't apply to structure types.  Together, that implies that top-level members cannot be structures, meaning there cannot be nested structures in input and output.

However, structures are not disallowed as top-level members of blocks (which in SPIR-V have a structure type).  The only way to make something flat in the following case is to make the block-structure member be decorated as flat, as the rules are clearly saying don't go deeper than that:

```
struct S {  vec4 v; };
in inName {  flat S s; };  // Allowed?  I think so, but that means a struct is Flat
```

Given that structures should not be disallowed here, I suspect the rule saying **Flat** only applies to scalars/vectors is the overly restrictive one, especially given how clearly the other rule mentions not decorating members of nested structures (as opposed to saying "no structures").

**Comment by ChrisGautier** [comment]
I agree that allowing **Flat** on structures makes sense. One thing though is that the _effective_ decorations of top-level structure members depend on both their own **OpMemberDecorate** and the **OpDecorate** of the object structure declaration. At the same time, the validation rules say:
```
 - At most one of Noperspective or Flat Decorations can be applied to the same object or member.
 - At most one of Patch, Centroid, or Sample Decorations can be applied to the same object or member.
 - At most one of RowMajor and ColMajor Decorations can be applied to a structure type.
 - At most one of Block and BufferBlock Decorations can be applied to a structure type.
```

So the following SPIR-V should be invalid:
```
               OpName %21 "S"
               OpMemberName %21 0 "a"
               OpName %23 "var"
               OpMemberDecorate %21 0 Noperspective
               OpDecorate %23 Flat
         %21 = OpTypeStruct %6
         %22 = OpTypePointer Input %21
         %23 = OpVariable %22 Input
```

**Comment by johnkslang** [comment]
@ChrisGautier, I agree that SPIR-V is invalid.

**Comment by johnkslang** [comment]
Khronos decided to remove the language 

> The object or member must be a scalar or vector of floating-point type or integer type.

to remove the inconsistency with the other requirement that it apply only to top-level block members, which can be structures.

This will appear in the next revision of SPIR-V.

**Comment by johnkslang** [comment]
Fixed in Revision 7 of SPIR-V 1.0.
